components/Dialogs/src/private/MDCAlertControllerView+Private.m

```diff
@@ -283,23 +292,28 @@ - (void)setAccessoryView:(UIView *)accessoryView {
   [self setNeedsLayout];
 }
 
-- (void)setButtonFont:(UIFont *)font {
-  _buttonFont = font;
-
-  [self updateButtonFont];
-}
-
 - (void)updateButtonFont {
-  UIFont *finalButtonFont = self.buttonFont ?: [[self class] buttonFontDefault];
-  if (self.mdc_adjustsFontForContentSizeCategory) {
-    if (self.adjustsFontForContentSizeCategoryWhenScaledFontIsUnavailable) {
-      finalButtonFont = [finalButtonFont
-          mdc_fontSizedForMaterialTextStyle:kTitleTextStyle
-                       scaledForDynamicType:self.mdc_adjustsFontForContentSizeCategory];
-    }
-  }
   for (MDCButton *button in self.actionManager.buttonsInActionOrder) {
-    [button setTitleFont:finalButtonFont forState:UIControlStateNormal];
+    if (self.mdc_adjustsFontForContentSizeCategory) {
+      UIFont *finalButtonFont = [MDCAlertControllerView buttonFontDefault];
+      if (button.enableTitleFontForState) {
+        finalButtonFont = [button titleFontForState:UIControlStateNormal] ?: finalButtonFont;
+      } else {
+        finalButtonFont = button.titleLabel.font ?: finalButtonFont;
+      }
+
+      if (self.adjustsFontForContentSizeCategoryWhenScaledFontIsUnavailable &&
+          (finalButtonFont.mdc_scalingCurve == nil)) {
+        finalButtonFont = [finalButtonFont
+            mdc_fontSizedForMaterialTextStyle:kTitleTextStyle
+                         scaledForDynamicType:self.mdc_adjustsFontForContentSizeCategory];
+      }
+      if (button.enableTitleFontForState) {
+        [button setTitleFont:finalButtonFont forState:UIControlStateNormal];
+      } else {
+        button.titleLabel.font = finalButtonFont;
+      }
+    }
```

Comment:
tip: as discussed offline: this logic may be valuable to pull out to its own PR.